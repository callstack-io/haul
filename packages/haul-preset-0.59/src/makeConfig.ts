import {
  Runtime,
  EnvOptions,
  BundleConfig,
  BundleConfigBuilder,
  ProjectConfig,
  NormalizedProjectConfig,
  NormalizedProjectConfigBuilder,
  DEFAULT_PORT,
  NormalizedBundleConfig,
} from '@haul-bundler/core';
import RamBundlePlugin from '@haul-bundler/ram-bundle-webpack-plugin';
import BasicBundleWebpackPlugin from '@haul-bundler/basic-bundle-webpack-plugin';
import get from 'lodash.get';
import webpack from 'webpack';
import path from 'path';
import getDefaultConfig from './defaultConfig';

function makeAbsolute(
  root: string,
  value: string | string[]
): string | string[] {
  const resolve = (v: string): string =>
    path.isAbsolute(v) ? v : path.join(root, v);

  if (typeof value === 'string') {
    return resolve(value);
  }

  return value.map(item => resolve(item));
}

export default function makeConfig(
  projectConfig: ProjectConfig
): NormalizedProjectConfigBuilder {
  return (runtime: Runtime, env: EnvOptions) => {
    const normalizedServerConfig = {
      port: get(projectConfig, 'server.port', DEFAULT_PORT),
      host: get(projectConfig, 'server.host', 'localhost'),
    };

    const normalizedBundleConfigs: {
      [bundleName: string]: NormalizedBundleConfig;
    } = {};
    const webpackConfigs: { [bundleName: string]: webpack.Configuration } = {};

    Object.keys(projectConfig.bundles).forEach(bundleName => {
      const bundleConfigBuilder: BundleConfig | BundleConfigBuilder =
        projectConfig.bundles[bundleName];
      const bundleConfig: BundleConfig =
        typeof bundleConfigBuilder === 'function'
          ? bundleConfigBuilder(env)
          : bundleConfigBuilder;

      // TODO: add `dll: boolean` flag and push DllPlugin
      // TODO: use dllDependencies and push DllReferencePlugin
      // TODO: use `library` and `libraryTarget` options when necessary
      const normalizedBundleConfig = {
        entry: bundleConfig.entry,
        type: bundleConfig.type || env.bundleType || 'basic-bundle',
        platform: bundleConfig.platform || env.platform,
        root: bundleConfig.root || env.root,
        dev: bundleConfig.dev || env.dev,
        assetsDest: bundleConfig.assetsDest || env.assetsDest || '',
        minify: bundleConfig.minify || Boolean(env.minify),
        minifyOptions: bundleConfig.minifyOptions || undefined,
        sourceMap: bundleConfig.sourceMap || false,
        dllDependencies: bundleConfig.dllDependencies || [],
        providesModuleNodeModules: bundleConfig.providesModuleNodeModules || [
          'react-native',
        ],
        hasteOptions: bundleConfig.hasteOptions || {},
      };

      // Make sure all entries are absolute.
      normalizedBundleConfig.entry = makeAbsolute(
        normalizedBundleConfig.root,
        normalizedBundleConfig.entry
      );

      let webpackConfig = getDefaultConfig(runtime, normalizedBundleConfig, {
        bundle: Boolean(env.bundle),
        ...normalizedServerConfig,
      });

      // Tweak bundle when creating static bundle
      if (env.bundle && normalizedBundleConfig.type === 'basic-bundle') {
        (webpackConfig.plugins as webpack.Plugin[]).push(
          new BasicBundleWebpackPlugin(
            Boolean(env.bundle),
            Boolean(normalizedBundleConfig.sourceMap)
          )
        );
      } else if (env.bundle) {
        (webpackConfig.plugins as webpack.Plugin[]).push(
          new RamBundlePlugin({
            minify: normalizedBundleConfig.minify,
            minifyOptions: normalizedBundleConfig.minifyOptions,
            sourceMap: Boolean(normalizedBundleConfig.sourceMap),
            indexRamBundle:
              normalizedBundleConfig.type === 'indexed-ram-bundle',
            singleBundleMode: env.singleBundleMode,
          })
        );
      }

      if (env.assetsDest) {
        webpackConfig.output!.path = path.isAbsolute(env.assetsDest)
          ? env.assetsDest
          : path.join(normalizedBundleConfig.root, env.assetsDest);
      }

      if (env.bundleOutput) {
        webpackConfig.output!.filename = path.isAbsolute(env.bundleOutput)
          ? path.relative(webpackConfig.output!.path!, env.bundleOutput)
          : path.relative(
              webpackConfig.output!.path!,
              path.join(normalizedBundleConfig.root, env.bundleOutput)
            );
      }

      if (env.sourcemapOutput) {
        webpackConfig.output!.sourceMapFilename = path.isAbsolute(
          env.sourcemapOutput
        )
          ? path.relative(webpackConfig.output!.path!, env.sourcemapOutput)
          : path.relative(
              webpackConfig.output!.path!,
              path.join(normalizedBundleConfig.root, env.sourcemapOutput)
            );
      }

      const { transform } = bundleConfig;
      if (transform) {
        webpackConfig = transform({
          bundleName,
          config: webpackConfig,
          env,
          runtime,
        });
      }

      normalizedBundleConfigs[bundleName] = normalizedBundleConfig;
      webpackConfigs[bundleName] = webpackConfig;
    });

    return {
      server: normalizedServerConfig,
      bundles: normalizedBundleConfigs,
      webpackConfigs,
    } as NormalizedProjectConfig;
  };
}
