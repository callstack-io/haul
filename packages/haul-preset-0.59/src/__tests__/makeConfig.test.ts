import makeConfig from '../makeConfig';
import withPolyfills from '../withPolyfills';
import { Runtime, ProjectConfig, EnvOptions } from '@haul-bundler/core';

function hasPlugin(webpackConfig: any, pluginName: string) {
  return webpackConfig.plugins.some(
    (p: any) => p.constructor.name === pluginName
  );
}

describe('makeConfig', () => {
  describe('with single bundle', () => {
    it('should create config for basic-bundle', () => {
      const projectConfig: ProjectConfig = {
        bundles: {
          index: {
            entry: withPolyfills('index.js'),
          },
        },
      };
      const runtime = new Runtime();
      const env: EnvOptions = {
        platform: 'ios',
        root: __dirname,
        dev: true,
        bundle: true,
        bundleType: 'basic-bundle',
        singleBundleMode: true,
      };
      const config = makeConfig(projectConfig)(runtime, env);
      expect(
        hasPlugin(config.webpackConfigs.index, 'WebpackBasicBundlePlugin')
      ).toBeTruthy();
      expect(config).toMatchSnapshot();
    });

    it('should create config for indexed-ram-bundle', () => {
      const projectConfig: ProjectConfig = {
        bundles: {
          index: {
            type: 'indexed-ram-bundle',
            entry: withPolyfills('index.js'),
          },
        },
      };
      const runtime = new Runtime();
      const env: EnvOptions = {
        platform: 'ios',
        root: __dirname,
        dev: true,
        bundle: true,
        bundleType: 'basic-bundle',
        singleBundleMode: true,
      };
      const config = makeConfig(projectConfig)(runtime, env);
      expect(
        hasPlugin(config.webpackConfigs.index, 'WebpackRamBundlePlugin')
      ).toBeTruthy();
      expect(config).toMatchSnapshot();
    });

    it('should create config for file-ram-bundle', () => {
      const projectConfig: ProjectConfig = {
        bundles: {
          index: {
            type: 'file-ram-bundle',
            entry: withPolyfills('index.js'),
          },
        },
      };
      const runtime = new Runtime();
      const env: EnvOptions = {
        platform: 'ios',
        root: __dirname,
        dev: true,
        bundle: true,
        bundleType: 'basic-bundle',
        singleBundleMode: true,
      };
      const config = makeConfig(projectConfig)(runtime, env);
      expect(
        hasPlugin(config.webpackConfigs.index, 'WebpackRamBundlePlugin')
      ).toBeTruthy();
      expect(config).toMatchSnapshot();
    });
  });

  it('should create config with external source maps', () => {
    const projectConfig: ProjectConfig = {
      bundles: {
        index: {
          sourceMap: true,
          entry: withPolyfills('index.js'),
        },
      },
    };
    const runtime = new Runtime();
    const env: EnvOptions = {
      platform: 'ios',
      root: __dirname,
      dev: true,
      bundleType: 'basic-bundle',
      singleBundleMode: true,
    };
    let config = makeConfig(projectConfig)(runtime, env);
    expect(
      hasPlugin(config.webpackConfigs.index, 'SourceMapDevToolPlugin')
    ).toBeTruthy();

    config = makeConfig(projectConfig)(runtime, {
      ...env,
      dev: false,
    });
    expect(
      hasPlugin(config.webpackConfigs.index, 'SourceMapDevToolPlugin')
    ).toBeTruthy();
  });

  it('should create config with inline source maps', () => {
    const projectConfig: ProjectConfig = {
      bundles: {
        index: {
          sourceMap: 'inline',
          entry: withPolyfills('index.js'),
        },
      },
    };
    const runtime = new Runtime();
    const env: EnvOptions = {
      platform: 'ios',
      root: __dirname,
      dev: true,
      bundleType: 'basic-bundle',
      singleBundleMode: true,
    };
    const config = makeConfig(projectConfig)(runtime, env);
    expect(
      hasPlugin(config.webpackConfigs.index, 'EvalSourceMapDevToolPlugin')
    ).toBeTruthy();
  });

  it('should create config for packager server', () => {
    const projectConfig: ProjectConfig = {
      bundles: {
        index: {
          entry: withPolyfills('index.js'),
        },
      },
    };
    const runtime = new Runtime();
    const env: EnvOptions = {
      platform: 'ios',
      root: __dirname,
      dev: true,
      bundleType: 'basic-bundle',
      singleBundleMode: true,
    };
    const config = makeConfig(projectConfig)(runtime, env);
    expect(
      hasPlugin(config.webpackConfigs.index, 'WebpackBasicBundlePlugin')
    ).toBeFalsy();
    expect(
      hasPlugin(config.webpackConfigs.index, 'WebpackRamBundlePlugin')
    ).toBeFalsy();
  });
});
