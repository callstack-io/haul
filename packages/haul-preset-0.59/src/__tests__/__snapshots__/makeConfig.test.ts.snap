// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeConfig with single bundle should create config for basic-bundle 1`] = `
Object {
  "bundles": Object {
    "index": Object {
      "assetsDest": "",
      "dev": true,
      "dllDependencies": Array [],
      "entry": Array [
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/console.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/error-guard.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Number.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/String.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es7.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/babelHelpers.js",
        "<<NODE_MODULE>>/react-native/Libraries/Core/InitializeCore.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__/index.js",
      ],
      "hasteOptions": Object {},
      "minify": false,
      "minifyOptions": undefined,
      "platform": "ios",
      "providesModuleNodeModules": Array [
        "react-native",
      ],
      "root": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
      "sourceMap": false,
      "type": "basic-bundle",
    },
  },
  "server": Object {
    "host": "localhost",
    "port": 8081,
  },
  "webpackConfigs": Object {
    "index": Object {
      "context": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
      "devtool": false,
      "entry": Array [
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/console.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/error-guard.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Number.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/String.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es7.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/babelHelpers.js",
        "<<NODE_MODULE>>/react-native/Libraries/Core/InitializeCore.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__/index.js",
      ],
      "mode": "development",
      "module": Object {
        "rules": Array [
          Object {
            "parser": Object {
              "requireEnsure": false,
            },
          },
          Object {
            "exclude": /node_modules\\(\\?!\\.\\*\\[\\\\/\\\\\\\\\\]\\(react\\|@react-navigation\\|@react-native-community\\|@expo\\|pretty-format\\|@haul-bundler\\|metro\\)\\)/,
            "test": /\\\\\\.js\\$/,
            "use": Array [
              Object {
                "loader": "<<NODE_MODULE>>/babel-loader/lib/index.js",
                "options": Object {
                  "cacheDirectory": true,
                  "extends": undefined,
                  "plugins": Array [
                    "<<REPLACED>>/packages/haul-core-legacy/build/utils/fixRequireIssues.js",
                  ],
                },
              },
            ],
          },
          Object {
            "test": /\\\\\\.\\(aac\\|aiff\\|bmp\\|caf\\|gif\\|html\\|jpeg\\|jpg\\|m4a\\|m4v\\|mov\\|mp3\\|mp4\\|mpeg\\|mpg\\|obj\\|otf\\|pdf\\|png\\|psd\\|svg\\|ttf\\|wav\\|webm\\|webp\\)\\$/,
            "use": Object {
              "loader": "<<REPLACED>>/packages/haul-core/build/webpack/loaders/assetLoader",
              "options": Object {
                "bundle": true,
                "platform": "ios",
                "root": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
                "runtime": Runtime {
                  "inspectorClient": undefined,
                  "logger": Logger {
                    "debug": [Function],
                    "done": [Function],
                    "enhanceWithColor": [Function],
                    "enhanceWithModifier": [Function],
                    "error": [Function],
                    "info": [Function],
                    "inspectorClient": undefined,
                    "print": [Function],
                    "warn": [Function],
                  },
                },
              },
            },
          },
        ],
      },
      "optimization": Object {
        "concatenateModules": true,
        "minimize": false,
        "minimizer": Array [
          TerserPlugin {
            "options": Object {
              "cache": true,
              "cacheKeys": [Function],
              "chunkFilter": [Function],
              "exclude": undefined,
              "extractComments": false,
              "include": undefined,
              "minify": undefined,
              "parallel": true,
              "sourceMap": true,
              "terserOptions": Object {
                "output": Object {
                  "comments": /\\^\\\\\\*\\*!\\|@preserve\\|@license\\|@cc_on/i,
                },
              },
              "test": /\\\\\\.\\(js\\|\\(js\\)\\?bundle\\)\\(\\$\\|\\\\\\?\\)/i,
              "warningsFilter": [Function],
            },
          },
        ],
        "namedModules": true,
      },
      "output": Object {
        "filename": "index.ios.bundle",
        "globalObject": "this",
        "path": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
        "publicPath": "http://localhost:8081/",
      },
      "plugins": Array [
        CaseSensitivePathsPlugin {
          "fsOperations": 0,
          "options": Object {},
          "pathCache": Object {},
          "primed": false,
        },
        DefinePlugin {
          "definitions": Object {
            "__DEV__": true,
            "process.env": Object {
              "NODE_ENV": "\\"development\\"",
            },
          },
        },
        LoaderOptionsPlugin {
          "options": Object {
            "debug": true,
            "minimize": false,
            "test": Object {
              "test": [Function],
            },
          },
        },
        WebpackBasicBundlePlugin {
          "bundle": true,
          "sourceMap": false,
        },
      ],
      "resolve": Object {
        "alias": Object {
          "react-proxy": "@zamotany/react-proxy",
        },
        "extensions": Array [
          ".ios.js",
          ".native.js",
          ".js",
        ],
        "mainFields": Array [
          "react-native",
          "browser",
          "main",
        ],
        "plugins": Array [
          HasteResolver {
            "options": Object {
              "directories": Array [
                "<<NODE_MODULE>>/react-native/Libraries",
              ],
              "hasteOptions": Object {},
            },
          },
          AssetResolver {
            "options": Object {
              "platform": "ios",
              "runtime": Runtime {
                "inspectorClient": undefined,
                "logger": Logger {
                  "debug": [Function],
                  "done": [Function],
                  "enhanceWithColor": [Function],
                  "enhanceWithModifier": [Function],
                  "error": [Function],
                  "info": [Function],
                  "inspectorClient": undefined,
                  "print": [Function],
                  "warn": [Function],
                },
              },
            },
          },
        ],
      },
      "stats": "verbose",
      "target": "webworker",
    },
  },
}
`;

exports[`makeConfig with single bundle should create config for file-ram-bundle 1`] = `
Object {
  "bundles": Object {
    "index": Object {
      "assetsDest": "",
      "dev": true,
      "dllDependencies": Array [],
      "entry": Array [
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/console.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/error-guard.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Number.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/String.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es7.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/babelHelpers.js",
        "<<NODE_MODULE>>/react-native/Libraries/Core/InitializeCore.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__/index.js",
      ],
      "hasteOptions": Object {},
      "minify": false,
      "minifyOptions": undefined,
      "platform": "ios",
      "providesModuleNodeModules": Array [
        "react-native",
      ],
      "root": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
      "sourceMap": false,
      "type": "file-ram-bundle",
    },
  },
  "server": Object {
    "host": "localhost",
    "port": 8081,
  },
  "webpackConfigs": Object {
    "index": Object {
      "context": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
      "devtool": false,
      "entry": Array [
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/console.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/error-guard.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Number.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/String.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es7.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/babelHelpers.js",
        "<<NODE_MODULE>>/react-native/Libraries/Core/InitializeCore.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__/index.js",
      ],
      "mode": "development",
      "module": Object {
        "rules": Array [
          Object {
            "parser": Object {
              "requireEnsure": false,
            },
          },
          Object {
            "exclude": /node_modules\\(\\?!\\.\\*\\[\\\\/\\\\\\\\\\]\\(react\\|@react-navigation\\|@react-native-community\\|@expo\\|pretty-format\\|@haul-bundler\\|metro\\)\\)/,
            "test": /\\\\\\.js\\$/,
            "use": Array [
              Object {
                "loader": "<<NODE_MODULE>>/babel-loader/lib/index.js",
                "options": Object {
                  "cacheDirectory": true,
                  "extends": undefined,
                  "plugins": Array [
                    "<<REPLACED>>/packages/haul-core-legacy/build/utils/fixRequireIssues.js",
                  ],
                },
              },
            ],
          },
          Object {
            "test": /\\\\\\.\\(aac\\|aiff\\|bmp\\|caf\\|gif\\|html\\|jpeg\\|jpg\\|m4a\\|m4v\\|mov\\|mp3\\|mp4\\|mpeg\\|mpg\\|obj\\|otf\\|pdf\\|png\\|psd\\|svg\\|ttf\\|wav\\|webm\\|webp\\)\\$/,
            "use": Object {
              "loader": "<<REPLACED>>/packages/haul-core/build/webpack/loaders/assetLoader",
              "options": Object {
                "bundle": true,
                "platform": "ios",
                "root": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
                "runtime": Runtime {
                  "inspectorClient": undefined,
                  "logger": Logger {
                    "debug": [Function],
                    "done": [Function],
                    "enhanceWithColor": [Function],
                    "enhanceWithModifier": [Function],
                    "error": [Function],
                    "info": [Function],
                    "inspectorClient": undefined,
                    "print": [Function],
                    "warn": [Function],
                  },
                },
              },
            },
          },
        ],
      },
      "optimization": Object {
        "concatenateModules": true,
        "minimize": false,
        "minimizer": Array [
          TerserPlugin {
            "options": Object {
              "cache": true,
              "cacheKeys": [Function],
              "chunkFilter": [Function],
              "exclude": undefined,
              "extractComments": false,
              "include": undefined,
              "minify": undefined,
              "parallel": true,
              "sourceMap": true,
              "terserOptions": Object {
                "output": Object {
                  "comments": /\\^\\\\\\*\\*!\\|@preserve\\|@license\\|@cc_on/i,
                },
              },
              "test": /\\\\\\.\\(js\\|\\(js\\)\\?bundle\\)\\(\\$\\|\\\\\\?\\)/i,
              "warningsFilter": [Function],
            },
          },
        ],
        "namedModules": true,
      },
      "output": Object {
        "filename": "index.ios.bundle",
        "globalObject": "this",
        "path": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
        "publicPath": "http://localhost:8081/",
      },
      "plugins": Array [
        CaseSensitivePathsPlugin {
          "fsOperations": 0,
          "options": Object {},
          "pathCache": Object {},
          "primed": false,
        },
        DefinePlugin {
          "definitions": Object {
            "__DEV__": true,
            "process.env": Object {
              "NODE_ENV": "\\"development\\"",
            },
          },
        },
        LoaderOptionsPlugin {
          "options": Object {
            "debug": true,
            "minimize": false,
            "test": Object {
              "test": [Function],
            },
          },
        },
        WebpackRamBundlePlugin {
          "bundleName": "index",
          "indexRamBundle": false,
          "minify": false,
          "minifyOptions": undefined,
          "modules": Array [],
          "name": "WebpackRamBundlePlugin",
          "preloadBundles": Array [],
          "singleBundleMode": true,
          "sourceMap": false,
        },
      ],
      "resolve": Object {
        "alias": Object {
          "react-proxy": "@zamotany/react-proxy",
        },
        "extensions": Array [
          ".ios.js",
          ".native.js",
          ".js",
        ],
        "mainFields": Array [
          "react-native",
          "browser",
          "main",
        ],
        "plugins": Array [
          HasteResolver {
            "options": Object {
              "directories": Array [
                "<<NODE_MODULE>>/react-native/Libraries",
              ],
              "hasteOptions": Object {},
            },
          },
          AssetResolver {
            "options": Object {
              "platform": "ios",
              "runtime": Runtime {
                "inspectorClient": undefined,
                "logger": Logger {
                  "debug": [Function],
                  "done": [Function],
                  "enhanceWithColor": [Function],
                  "enhanceWithModifier": [Function],
                  "error": [Function],
                  "info": [Function],
                  "inspectorClient": undefined,
                  "print": [Function],
                  "warn": [Function],
                },
              },
            },
          },
        ],
      },
      "stats": "verbose",
      "target": "webworker",
    },
  },
}
`;

exports[`makeConfig with single bundle should create config for indexed-ram-bundle 1`] = `
Object {
  "bundles": Object {
    "index": Object {
      "assetsDest": "",
      "dev": true,
      "dllDependencies": Array [],
      "entry": Array [
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/console.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/error-guard.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Number.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/String.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es7.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/babelHelpers.js",
        "<<NODE_MODULE>>/react-native/Libraries/Core/InitializeCore.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__/index.js",
      ],
      "hasteOptions": Object {},
      "minify": false,
      "minifyOptions": undefined,
      "platform": "ios",
      "providesModuleNodeModules": Array [
        "react-native",
      ],
      "root": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
      "sourceMap": false,
      "type": "indexed-ram-bundle",
    },
  },
  "server": Object {
    "host": "localhost",
    "port": 8081,
  },
  "webpackConfigs": Object {
    "index": Object {
      "context": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
      "devtool": false,
      "entry": Array [
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/console.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/error-guard.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Number.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/String.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.prototype.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Array.es6.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/Object.es7.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/vendor/polyfills/babelHelpers.js",
        "<<NODE_MODULE>>/react-native/Libraries/Core/InitializeCore.js",
        "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__/index.js",
      ],
      "mode": "development",
      "module": Object {
        "rules": Array [
          Object {
            "parser": Object {
              "requireEnsure": false,
            },
          },
          Object {
            "exclude": /node_modules\\(\\?!\\.\\*\\[\\\\/\\\\\\\\\\]\\(react\\|@react-navigation\\|@react-native-community\\|@expo\\|pretty-format\\|@haul-bundler\\|metro\\)\\)/,
            "test": /\\\\\\.js\\$/,
            "use": Array [
              Object {
                "loader": "<<NODE_MODULE>>/babel-loader/lib/index.js",
                "options": Object {
                  "cacheDirectory": true,
                  "extends": undefined,
                  "plugins": Array [
                    "<<REPLACED>>/packages/haul-core-legacy/build/utils/fixRequireIssues.js",
                  ],
                },
              },
            ],
          },
          Object {
            "test": /\\\\\\.\\(aac\\|aiff\\|bmp\\|caf\\|gif\\|html\\|jpeg\\|jpg\\|m4a\\|m4v\\|mov\\|mp3\\|mp4\\|mpeg\\|mpg\\|obj\\|otf\\|pdf\\|png\\|psd\\|svg\\|ttf\\|wav\\|webm\\|webp\\)\\$/,
            "use": Object {
              "loader": "<<REPLACED>>/packages/haul-core/build/webpack/loaders/assetLoader",
              "options": Object {
                "bundle": true,
                "platform": "ios",
                "root": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
                "runtime": Runtime {
                  "inspectorClient": undefined,
                  "logger": Logger {
                    "debug": [Function],
                    "done": [Function],
                    "enhanceWithColor": [Function],
                    "enhanceWithModifier": [Function],
                    "error": [Function],
                    "info": [Function],
                    "inspectorClient": undefined,
                    "print": [Function],
                    "warn": [Function],
                  },
                },
              },
            },
          },
        ],
      },
      "optimization": Object {
        "concatenateModules": true,
        "minimize": false,
        "minimizer": Array [
          TerserPlugin {
            "options": Object {
              "cache": true,
              "cacheKeys": [Function],
              "chunkFilter": [Function],
              "exclude": undefined,
              "extractComments": false,
              "include": undefined,
              "minify": undefined,
              "parallel": true,
              "sourceMap": true,
              "terserOptions": Object {
                "output": Object {
                  "comments": /\\^\\\\\\*\\*!\\|@preserve\\|@license\\|@cc_on/i,
                },
              },
              "test": /\\\\\\.\\(js\\|\\(js\\)\\?bundle\\)\\(\\$\\|\\\\\\?\\)/i,
              "warningsFilter": [Function],
            },
          },
        ],
        "namedModules": true,
      },
      "output": Object {
        "filename": "index.ios.bundle",
        "globalObject": "this",
        "path": "<<REPLACED>>/packages/haul-preset-0.59/src/__tests__",
        "publicPath": "http://localhost:8081/",
      },
      "plugins": Array [
        CaseSensitivePathsPlugin {
          "fsOperations": 0,
          "options": Object {},
          "pathCache": Object {},
          "primed": false,
        },
        DefinePlugin {
          "definitions": Object {
            "__DEV__": true,
            "process.env": Object {
              "NODE_ENV": "\\"development\\"",
            },
          },
        },
        LoaderOptionsPlugin {
          "options": Object {
            "debug": true,
            "minimize": false,
            "test": Object {
              "test": [Function],
            },
          },
        },
        WebpackRamBundlePlugin {
          "bundleName": "index",
          "indexRamBundle": true,
          "minify": false,
          "minifyOptions": undefined,
          "modules": Array [],
          "name": "WebpackRamBundlePlugin",
          "preloadBundles": Array [],
          "singleBundleMode": true,
          "sourceMap": false,
        },
      ],
      "resolve": Object {
        "alias": Object {
          "react-proxy": "@zamotany/react-proxy",
        },
        "extensions": Array [
          ".ios.js",
          ".native.js",
          ".js",
        ],
        "mainFields": Array [
          "react-native",
          "browser",
          "main",
        ],
        "plugins": Array [
          HasteResolver {
            "options": Object {
              "directories": Array [
                "<<NODE_MODULE>>/react-native/Libraries",
              ],
              "hasteOptions": Object {},
            },
          },
          AssetResolver {
            "options": Object {
              "platform": "ios",
              "runtime": Runtime {
                "inspectorClient": undefined,
                "logger": Logger {
                  "debug": [Function],
                  "done": [Function],
                  "enhanceWithColor": [Function],
                  "enhanceWithModifier": [Function],
                  "error": [Function],
                  "info": [Function],
                  "inspectorClient": undefined,
                  "print": [Function],
                  "warn": [Function],
                },
              },
            },
          },
        ],
      },
      "stats": "verbose",
      "target": "webworker",
    },
  },
}
`;
